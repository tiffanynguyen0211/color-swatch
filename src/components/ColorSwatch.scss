@use "../scss/variables.scss";

.ColorSwatch__ColorList {
  display: grid;
  overflow: auto;
  height: 300px;
  margin: 0 variables.$spacing-unit;
  grid-auto-rows: 40px;
  grid-template-columns: repeat(auto-fit, minmax(65px, 1fr));
}

.ColorSwatch__Display {
  width: 50%;
  margin-right: variables.$spacing-unit;
  margin-left: variables.$spacing-unit;

  &::before {
    padding-top: 50%;
    content: "";
    float: left;
  }
}

.ColorSwatch__Container {
  width: 100vw;
}

.ColorSwatch__DisplayContainer {
  display: flex;
  width: 100%;
  padding-bottom: 30px;
}

.ColorSwatch__Button {
  margin: 5px;
  box-shadow: 0 0 5px 3px rgb(0 0 0 / 14%);
  transition: transform 1s;

  &:hover,
  &:focus {
    box-shadow: 0 0 5px 3px rgb(0 0 0 / 14%);
    transform: scale(1.075);
  }
}

.ColorSwatch__Button--reset {
  margin: 0 variables.$spacing-unit 30px;
}
