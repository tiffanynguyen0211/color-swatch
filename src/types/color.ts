export type RgbColor = {
    red: number; // [0, 255]
    green: number; // [0, 255]
    blue: number; // [0, 255]
};

export type HslColor = {
    hue: number; // [0, 360] degrees
    saturation: number; // [0, 100] percentage
    lightness: number; // [0, 100] percentage
};

export type ColorListItem = {
    kind: 'rgb' | 'hsl';
    components: RgbColor | HslColor;
};

export type ColorList = ColorListItem[];
