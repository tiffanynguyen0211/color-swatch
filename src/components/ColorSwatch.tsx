import './ColorSwatch.scss';

import { useEffect, useState } from 'react';

import { ColorListItem } from '../types/color';
import { colorToRBGValue } from '../utils/colorValue';
import { CopyToClipboard } from './CopytoClipboard';

const ColorPanel: React.FC<
    {
        onClick: (value: ColorListItem) => void;
    } & ColorListItem
> = ({ components, kind, onClick }) => (
    <button
        onClick={() => onClick({ kind, components })}
        className="ColorSwatch__Button"
    >
        <span>
            <div
                style={{
                    backgroundColor: colorToRBGValue({ kind, components }),
                    padding: '15px',
                }}
            />
        </span>
    </button>
);

export const ColorSwatch: React.FC<{
    colorsData?: ColorListItem[] | null;
    onReset: () => void;
}> = ({ colorsData, onReset }) => {
    const [selected, setSelected] = useState<ColorListItem | null>(null);

    useEffect(() => {
        setSelected(null);
        if (colorsData && colorsData.length > 0) setSelected(colorsData[0]);
    }, [colorsData]);

    return (
        <div className="ColorSwatch__Container">
            <div className="ColorSwatch__DisplayContainer">
                {colorsData && colorsData.length > 0 && selected ? (
                    <>
                        <div
                            className="ColorSwatch__Display"
                            style={{
                                backgroundColor: colorToRBGValue(selected),
                            }}
                        ></div>
                        <CopyToClipboard copyText={colorToRBGValue(selected)} />
                    </>
                ) : (
                    // TODO add loading skeleton
                    <div style={{ margin: 'auto' }}>Loading...</div>
                )}
            </div>

            <button
                className="Button ColorSwatch__Button--reset"
                onClick={onReset}
            >
                Reload Color
            </button>
            <div className="ColorSwatch__ColorList">
                {colorsData &&
                    colorsData.length > 0 &&
                    colorsData.map((eachData, index) => (
                        <ColorPanel
                            key={`${eachData.kind}-${index}`}
                            {...eachData}
                            onClick={(currentColor) => {
                                setSelected(currentColor);
                            }}
                        />
                    ))}
                {colorsData && colorsData.length === 0 && (
                    <p>No available colors</p>
                )}
                {/* TODO add loading skeleton */}
                {!colorsData && (
                    <div style={{ margin: 'auto' }}>Loading...</div>
                )}
            </div>
        </div>
    );
};

declare global {
    interface Window {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        clipboardData: any;
    }
}
