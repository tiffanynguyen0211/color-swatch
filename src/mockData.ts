import { ColorListItem } from './types/color';

export const mockData: ColorListItem[] = [
    { kind: 'rgb', components: { red: 206, green: 83, blue: 241 } },
    { kind: 'rgb', components: { red: 146, green: 210, blue: 132 } },
    { kind: 'rgb', components: { red: 245, green: 124, blue: 221 } },
    { kind: 'rgb', components: { red: 55, green: 241, blue: 76 } },
    { kind: 'rgb', components: { red: 181, green: 20, blue: 205 } },
    { kind: 'rgb', components: { red: 206, green: 83, blue: 241 } },
    { kind: 'rgb', components: { red: 146, green: 210, blue: 132 } },
    { kind: 'rgb', components: { red: 245, green: 124, blue: 221 } },
    { kind: 'rgb', components: { red: 55, green: 241, blue: 76 } },
    { kind: 'rgb', components: { red: 181, green: 20, blue: 205 } },
    { kind: 'hsl', components: { hue: 0, saturation: 100, lightness: 50 } },
    { kind: 'hsl', components: { hue: 180, saturation: 50, lightness: 50 } },
];
