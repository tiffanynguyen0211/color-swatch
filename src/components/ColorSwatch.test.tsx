import axe from 'axe-core';
import { ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ReactElement } from 'react';

import { mountToDoc } from '../../.jest/jest-setup';
import { mockData } from '../mockData';
import { ColorListItem } from '../types/color';
import { ColorSwatch } from './ColorSwatch';

const setup = (
    render: (x: ReactElement) => ReactWrapper | ShallowWrapper,
    props: JSX.IntrinsicAttributes,
) => {
    const defaultProps: {
        colorsData: ColorListItem[] | null;
        onReset: () => void;
    } = {
        colorsData: mockData,
        onReset: jest.fn(),
    };
    const onClickSpy = jest.fn();

    const component = render(<ColorSwatch {...defaultProps} {...props} />);

    return {
        actual: component,
        onClickSpy,
    };
};

describe('ColorSwatch', () => {
    it('should match snapshot on first render', () => {
        const { actual } = setup(shallow, {});
        expect(toJson(actual)).toMatchSnapshot();
    });

    it('should match snapshot on empty data render', () => {
        const { actual } = setup(shallow, {});
        actual.setProps({ colorsData: null });
        expect(toJson(actual)).toMatchSnapshot();
    });

    it('has no accessibility violations', () => {
        const { actual } = setup(mountToDoc, {});
        const componentNode = actual.getDOMNode();

        axe.run(componentNode, (err, { violations }) => {
            expect(err).toBe(null);
            expect(violations).toHaveLength(0);
        });
    });
});
