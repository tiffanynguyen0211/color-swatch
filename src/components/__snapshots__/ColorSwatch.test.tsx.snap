// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorSwatch should match snapshot on empty data render 1`] = `
<div
  className="ColorSwatch__Container"
>
  <div
    className="ColorSwatch__DisplayContainer"
  >
    <div
      style={
        Object {
          "margin": "auto",
        }
      }
    >
      Loading...
    </div>
  </div>
  <button
    className="Button ColorSwatch__Button--reset"
    onClick={[MockFunction]}
  >
    Reload Color
  </button>
  <div
    className="ColorSwatch__ColorList"
  >
    <div
      style={
        Object {
          "margin": "auto",
        }
      }
    >
      Loading...
    </div>
  </div>
</div>
`;

exports[`ColorSwatch should match snapshot on first render 1`] = `
<div
  className="ColorSwatch__Container"
>
  <div
    className="ColorSwatch__DisplayContainer"
  >
    <div
      style={
        Object {
          "margin": "auto",
        }
      }
    >
      Loading...
    </div>
  </div>
  <button
    className="Button ColorSwatch__Button--reset"
    onClick={[MockFunction]}
  >
    Reload Color
  </button>
  <div
    className="ColorSwatch__ColorList"
  >
    <ColorPanel
      components={
        Object {
          "blue": 241,
          "green": 83,
          "red": 206,
        }
      }
      key="rgb-0"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "blue": 132,
          "green": 210,
          "red": 146,
        }
      }
      key="rgb-1"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "blue": 221,
          "green": 124,
          "red": 245,
        }
      }
      key="rgb-2"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "blue": 76,
          "green": 241,
          "red": 55,
        }
      }
      key="rgb-3"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "blue": 205,
          "green": 20,
          "red": 181,
        }
      }
      key="rgb-4"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "blue": 241,
          "green": 83,
          "red": 206,
        }
      }
      key="rgb-5"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "blue": 132,
          "green": 210,
          "red": 146,
        }
      }
      key="rgb-6"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "blue": 221,
          "green": 124,
          "red": 245,
        }
      }
      key="rgb-7"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "blue": 76,
          "green": 241,
          "red": 55,
        }
      }
      key="rgb-8"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "blue": 205,
          "green": 20,
          "red": 181,
        }
      }
      key="rgb-9"
      kind="rgb"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "hue": 0,
          "lightness": 50,
          "saturation": 100,
        }
      }
      key="hsl-10"
      kind="hsl"
      onClick={[Function]}
    />
    <ColorPanel
      components={
        Object {
          "hue": 180,
          "lightness": 50,
          "saturation": 50,
        }
      }
      key="hsl-11"
      kind="hsl"
      onClick={[Function]}
    />
  </div>
</div>
`;
